### PR TITLE
WIP: cluster:swarm runner

### PR DESCRIPTION
**Pull request by jimpick**
I have an automated Ansible setup for Docker swarm in the aws-ansible branch (needs just a bit more cleanup). I first tested it with a shell script, but then I realized that it wouldn't be too much work to modify the local:docker runner to talk to the swarm API.

Here's a demo running the smlbench2 2-container bitswap test across 2 different EC2 vms in the Docker swarm cluster:

tg-docker-swarm-runner

https://github.com/ipfs/testground/blob/aws-ansible/pkg/runner/cluster_swarm.go

It's not quite done yet ... I've hard-coded the image (pulling from ECR), the REDIS_HOST environment variable, and the network. We'll want to upload the image to ECR after building it before creating the service.

-------------- 

Not complete yet ... some hard-coded values. See demo of it running in #110 (along with notes about what is still missing)

I just hacked this together to test the ansible setup I've been prototyping in the aws-ansible branch:

https://github.com/ipfs/testground/tree/aws-ansible

(tested with the smlbench2 test, which isn't on master yet ... that's going to be the basis of #94)

### CODE REVIEW

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ var AllBuilders = []build.Builder{
 var AllRunners = []runner.Runner{
 	&runner.LocalDockerRunner{},
 	&runner.LocalExecutableRunner{},
+	&runner.ClusterSwarmRunner{},
```
Comment (daviddias) [inline pkg/engine/engine.go]:
```suggestion
	&runner.ClusterDockerSwarmRunner{},
```

Comment (daviddias) [reply pkg/engine/engine.go]: I'm not entirely sure if naming yet another Runner just for Docker Swarm is the best path. It probably worth to just have Docker Swarm as the default Docker Runner

Comment (daviddias) [reply pkg/engine/engine.go]: Or perhaps this is me just getting confused by seeing Docker Swarm being called a Runner (given that in the end it is just the same thing running in a Docker container, but managed by Docker Swarm)

File: pkg/runner/cluster_swarm.go [hunk 2]
```diff
@@ -1 +1,162 @@
...
+	// Start service
+	// name := fmt.Sprintf("tg-%s-%s-%s", input.TestPlan.Name, testcase.Name, input.ID)
+	name := fmt.Sprintf("tg-%s-%s", input.TestPlan.Name, testcase.Name)
+
+	log.Infow("creating service", "name", name)
+
+  // docker service create --replicas 1 --name helloworld alpine ping docker.com
+
+  replicas := uint64(input.Instances)
+  spec := swarm.ServiceSpec{
```
Comment (daviddias) [inline pkg/runner/cluster_swarm.go]: Can I get a clarification of this way of specing the service vs. using docker-compose.yml? (https://docs.docker.com/compose/)

File: pkg/engine/engine.go (hunk 1 above)
Comment (raulk) [reply pkg/engine/engine.go]: I think calling this `ClusterSwarmRunner` is fine. I was using the convention `scope:technology`, e.g. `local:exec`, `local:docker`, etc. to name runners. We can also rename "runners" to "schedulers" if we feel that's more accurate.

File: pkg/runner/cluster_swarm.go (hunk 2 above)
Comment (raulk) [reply pkg/runner/cluster_swarm.go]: Now that we've decided to create a temporary overlay network per run, we _could_ use Docker Compose to create both the network and the service in one go. However, IIRC Docker Compose is really a CLI tool written in Python that transforms a YAML into docker calls. We cannot use it from Golang :-(

**Comment by raulk** [comment]
Taking this to land it on top of @jimpick's work on the infra side of things.

**Comment by raulk** [comment]
Closing this PR in favour of #126. Thanks for the head start here, @jimpick! Let's land it!

**Pull request closed without merging**